src/profilers/profile-serializer.ts

```diff
@@ -0,0 +1,145 @@
+import {perftools} from '../profile';
+import {getIndexOrAdd} from '../util';
+import {WallProfile, WallProfileNode} from '../v8-types';
+// A stack of function UIDs.
+type Stack = Array<number>;
+
+// Helper class used to convert v8 profile into profile format used by
+// Stackdriver Profiler.
+class ProfileSerializer {
+  private samples: Array<perftools.profiles.Sample>;
+  private locations: Array<perftools.profiles.Location>;
+  private functions: Array<perftools.profiles.Function>;
+  private locationMap: Map<number, perftools.profiles.Location>;
+  private functionMap: Map<number, perftools.profiles.Function>;
+  private strings: string[];
+  private samplingInterval: number;  // microseconds, defaults to 1000us
+
+  constructor() {
+    this.reset();
+  }
+
+  // Converts profile returned by v8 wall profiler to profile with format used
+  // by Stackdriver Profilers.
+  serializeWallProfile(prof: WallProfile, sampleInterval: number):
+      perftools.profiles.IProfile {
+    this.reset();
+    this.samplingInterval = sampleInterval;
+
+    let sampleValueType = this.sampleValueType();
+    let timeValueType = this.timeValueType();
+
+    this.serializeNode(prof.topDownRoot, []);
+
+    return {
+      sampleType: [sampleValueType, timeValueType],
+      sample: this.samples,
+      location: this.locations,
+      function: this.functions,
+      stringTable: this.strings,
+      // opt drop_frames
+      // opt keep_frames
+      timeNanos: 1000 * 1000 * prof.endTime,                         // Nanos
+      durationNanos: 1000 * 1000 * (prof.endTime - prof.startTime),  // Nanos
+
+      periodType: timeValueType,
+      period: this.samplingInterval
+    };
+  }
+
+  // clears fields used to build a profile.
+  private reset() {
```

Comment:
Looking at the implementation, it seems that the objects of this class are meant to be instantiated only for the duration of the `serializeWallProfile` function, after which the object, and state it holds are no longer necessary.

The pattern of holding state in a temporary short lived object, along with a `reset` function here doesn't seem like the best idea to me.

Instead, I think it would make sense for `serializeWallProfile` to be a top-level function that creates a closure with the necessary state. All the other private functions in the class can be nested functions on `serializeWallProfile`, hence getting access to the closure state. WDYT?

Reply:
That makes more sense to me, too. Will change.